File: app/src/main/java/com/example/appa/ui/SettingsFragment.java
package com.example.appa.ui;

import android.os.Bundle;

import androidx.fragment.app.Fragment;
import androidx.preference.ListPreference;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;

import com.example.appa.R;

public class SettingsFragment extends PreferenceFragmentCompat {

    public static Fragment newInstance(String rootKey) {
        SettingsFragment fragment = new SettingsFragment();
        Bundle bundle = new Bundle();
        bundle.putString(PreferenceFragmentCompat.ARG_PREFERENCE_ROOT, rootKey);
        fragment.setArguments(bundle);
        return fragment;
    }


    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        setPreferencesFromResource(R.xml.preferences, rootKey);

        switch (rootKey) {
            case "preference_appearance":
                onCreateAppearancePreferences();
                break;
            case "preference_others":
                break;
        }
    }

    private void onCreateAppearancePreferences() {

        final ListPreference themePreference = (ListPreference) findPreference("preference_theme");
        themePreference.setSummary(themePreference.getEntry());
        themePreference.setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
            @Override
            public boolean onPreferenceChange(Preference preference, Object newValue) {
                int indexOfValue = themePreference.findIndexOfValue(String.valueOf(newValue));
                themePreference.setSummary(indexOfValue >= 0 ? themePreference.getEntries()[indexOfValue] : null);
                return true;
            }
        });
    }

    @Override
    public void onResume() {
        super.onResume();

        String rootKey = getArguments().getString(PreferenceFragmentCompat.ARG_PREFERENCE_ROOT);
        getActivity().setTitle(findPreference(rootKey).getTitle());
    }

}

File: app/src/main/java/com/example/appa/ui/settings/SettingsFragment.kt
package com.example.appa.ui.settings

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.preference.*
import com.example.appa.R
import com.example.appa.ui.MainActivity

class SettingsFragment : PreferenceFragmentCompat() {


    // private val preferences: SharedPreferences = by inject()
    // activity.getSharedPreferences("preferences", Context.MODE_PRIVATE)
    //private val sharedpref = PreferenceManager.getDefaultSharedPreferences(context)




/*
    override fun onResume() {
        super.onResume()
        //preferenceScreen.sharedPreferences.registerOnSharedPreferenceChangeListener(this)
        PreferenceManager.getDefaultSharedPreferences(context)
                .registerOnSharedPreferenceChangeListener(this)
    }

    override fun onPause() {
        super.onPause()
        //preferenceScreen.sharedPreferences.unregisterOnSharedPreferenceChangeListener(this)
        PreferenceManager.getDefaultSharedPreferences(context)
                .unregisterOnSharedPreferenceChangeListener(this)
    }
    */

    companion object{
        fun newInstance(rootkey: String):SettingsFragment{
            var settingsFragment = SettingsFragment()
            var bundle = Bundle()
            bundle.putString(PreferenceFragmentCompat.ARG_PREFERENCE_ROOT, rootkey)
            settingsFragment.arguments = bundle
            return SettingsFragment()
        }
    }

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        setPreferencesFromResource(R.xml.root_preferences, rootKey)

        // Theme preference
        val theme = findPreference<ListPreference>("theme") as ListPreference
        theme.onPreferenceChangeListener = Preference.OnPreferenceChangeListener{preference, newValue ->
            val id = newValue as String
            ThemeSetting.setDefaultNightMode(ThemeEnum.idOf(id))
            val themePref = PreferenceManager.getDefaultSharedPreferences(context)
            val themeId = themePref.getString("theme", "")
            if(id != themeId){
                val contrastPref = PreferenceManager.getDefaultSharedPreferences(context)
                if(id.equals("1")) { // light theme
                    if (contrastPref.getBoolean("contrast", false) == false) {
                        themePref.edit().putString("style", "AppTheme_Light").apply()
                    } else {
                        themePref.edit().putString("style", "App_Theme_Light_HC").apply()
                    }
                }
                else if(id.equals("2")) { // Dark theme
                    if (contrastPref.getBoolean("contrast", false) == false) {
                        themePref.edit().putString("style", "AppTheme_Dark").apply()
                    } else {
                        themePref.edit().putString("style", "AppTheme_Dark_HC").apply()
                    }
                }
            }
            reload()
            true
        }

        val contrast = findPreference<SwitchPreference>("contrast") as SwitchPreference
        contrast.onPreferenceChangeListener = Preference.OnPreferenceChangeListener{preference, newValue ->
            val contrast = newValue as Boolean
            val contrastPref = PreferenceManager.getDefaultSharedPreferences(context)
            val constastId = contrastPref.getBoolean("contrast", false)
            if(contrast != constastId){
                if(contrast == false){ // not HC
                    if(contrastPref.getString("theme", "").equals("1")){
                        contrastPref.edit().putString("style", "AppTheme_Light").apply()
                    } else {
                        contrastPref.edit().putString("style", "AppTheme_Dark").apply()
                    }
                }
                else { // HC
                    if(contrastPref.getString("theme", "").equals("1")) {
                        contrastPref.edit().putString("style", "AppTheme_Light_HC").apply()
                    } else {
                        contrastPref.edit().putString("style", "AppTheme_Dark_HC").apply()
                    }
                }
                contrastPref.edit().putBoolean("contrast", contrast).apply()
                Log.d(tag, "This is for debug: " + contrast)
                val style = contrastPref.getString("style", "")
                Log.d(tag,"this is style: " + style)
            }
            startActivity(Intent(context, MainActivity::class.java))
            true
        }
    }

    private fun reload() {
        getParentFragmentManager().beginTransaction().remove(this).commit()
    }


/*
    private fun theme() {
        findPreference<ListPreference>("theme")?.apply {
            setOnPreferenceChangeListener{ preference, newValue ->
                val id = newValue as String
                ThemeSetting.setDefaultNightMode(ThemeEnum.idOf(id))
                true
            }
        }

    }*/



}

